### PR TITLE
feat: [CDS-74523]: add validations for remote input set selection across multiple panels

### PR DESCRIPTION
**Pull request by Eric761**
### Summary

Customer does not require feature branch selection at this time. All we need to do now is improve our validation and error handling for remote Input set selection.

Ticket: https://harness.atlassian.net/browse/CDS-74523
Parent Ticket: https://harness.atlassian.net/browse/CDS-71997

- Every time a `remote InputSet` is chosen, we validate it and display a spinner while doing so.
- If not found in current git context :
     - Appropriate error message will be displayed
     - Checkbox will be unchecked and disabled. 
- If found in the current git context, check icon will be displayed.
- Only after all validations have been successful, Apply Input Set button will be enabled.

#### Screenshots

Screenshot 2023-07-28 at 1 48 32 PM

#### Demo Video

https://github.com/harness/harness-core-ui/assets/73639897/3e1ae47f-6950-479d-b43c-b1d16933ffe1

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Review by vikrant-harness — DISMISSED** [review]

**Review by sunnykesh-harness — DISMISSED** [review]

File: src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.module.scss [hunk 1]
```diff
@@ -107,9 +107,8 @@
 }
 
 .popover-container {
-  min-width: 530px !important;
-  max-width: 700px !important;
-  max-height: 650px;
+  width: 550px !important;
```
Comment (rameshx) [inline src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.module.scss:110]: let's increase this to 600

File: src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.tsx [hunk 2]
```diff
@@ -189,14 +196,69 @@ export function InputSetSelector({
     }
   }, [invalidInputSetReferences])
 
+  const anyInputSetInLoadingState = useMemo(() => {
+    let anyInputSetLoading = false
+    for (const inputSetErrorMetaData of inputSetErrorMap.values()) {
```
Comment (rameshx) [inline src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.tsx:201]: use `Array.prototype.some`

File: src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.tsx [hunk 3]
```diff
@@ -189,14 +196,69 @@ export function InputSetSelector({
...
+      })
+      if (response?.status === 'ERROR') {
+        inputSetContainsError = true
+        inputSetBranch = get(response, 'metadata.branch', '')
+      }
+    } catch (err) {
+      inputSetContainsError = true
+    } finally {
+      setSelectedInputSets((prevSelectedInputSets: InputSetValue[]) =>
+        removeInvalidInputSet(prevSelectedInputSets, inputSetId, inputSetContainsError)
```
Comment (rameshx) [inline src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.tsx:234]: If `inputSetContainsError` is false then directly return the previous state. Only call `removeInvalidInputSet` when removal is necessary. After this, `removeInvalidInputSet` doesn't need the arg `inputSetContainsError`.

File: src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.tsx [hunk 4]
```diff
@@ -189,14 +196,69 @@ export function InputSetSelector({
...
+      )
+      setSelectedInputSetsContext?.((prevSelectedInputSets?: InputSetValue[]) => {
+        if (prevSelectedInputSets)
+          return removeInvalidInputSet(prevSelectedInputSets, inputSetId, inputSetContainsError)
+      })
```
Comment (rameshx) [inline src/modules/70-pipeline/components/InputSetSelector/InputSetSelector.tsx:239]: same as above

File: src/modules/70-pipeline/components/InputSetSelector/utils.ts [hunk 5]
```diff
@@ -62,4 +68,15 @@ export const getInputSetExpressionValue = (expressionVal: string): InputSetValue
...
 
+export const removeInvalidInputSet = (
+  selectedInputSets: InputSetValue[],
+  inputSetId: string,
+  inputSetContainsError: boolean
+): InputSetValue[] => {
+  const clonedInputSets = clone(selectedInputSets)
+  const removedItem = clonedInputSets.filter(set => set.value === inputSetId)[0]
+  if (removedItem && inputSetContainsError) clonedInputSets.splice(clonedInputSets.indexOf(removedItem), 1)
+  return clonedInputSets
```
Comment (rameshx) [inline src/modules/70-pipeline/components/InputSetSelector/utils.ts:79]: Optimize this to remove the item in one loop.

**Review by rameshx — CHANGES_REQUESTED** [review]

**Review by satyam-harness — APPROVED** [review]

File: src/modules/70-pipeline/components/InputSetSelector/utils.ts (hunk 5 above)
Comment (Eric761) [reply src/modules/70-pipeline/components/InputSetSelector/utils.ts]: done

File: src/modules/70-pipeline/components/InputSetSelector/utils.ts [hunk 6]
```diff
@@ -68,14 +68,9 @@ export const getInputSetExpressionValue = (expressionVal: string): InputSetValue
...
-export const removeInvalidInputSet = (
-  selectedInputSets: InputSetValue[],
-  inputSetId: string,
-  inputSetContainsError: boolean
-): InputSetValue[] => {
+export const removeInvalidInputSet = (selectedInputSets: InputSetValue[], inputSetId: string): InputSetValue[] => {
   const clonedInputSets = clone(selectedInputSets)
-  const removedItem = clonedInputSets.filter(set => set.value === inputSetId)[0]
-  if (removedItem && inputSetContainsError) clonedInputSets.splice(clonedInputSets.indexOf(removedItem), 1)
+  remove(clonedInputSets, set => set.value === inputSetId)
```
Comment (rameshx) [inline src/modules/70-pipeline/components/InputSetSelector/utils.ts:73]: `Array.prototype.filter` can be used to do this in one loop, but this is fine and much better than the previous code.

**Review by rameshx — APPROVED** [review]

**Pull request merged**